File: p1_python演習/my_answer/ex14.py
import numpy as np
docs = [["リンゴ", "リンゴ"], ["リンゴ", "レモン"], ["レモン", "ミカン"]]
terms = ["リンゴ", "レモン", "ミカン"]

def idf(term, docs):
    bunbo = 0
    bunshi = len(docs)
    for doc in docs:
        if term in doc:
            bunbo += 1
    
    return np.log10(bunshi / bunbo) + 1

for x in terms:
    print("tf(%s) = %f" % (x, idf(x, docs)))

File: p1_python演習/cpp_answer/ex14.cpp
#include<iostream>
#include<string>
#include<vector>
#include<set>
#include<fstream>
#include<cmath>

double idf(const std::string& term, const std::vector<std::vector<std::string>>& docs) {
    int df = 0;
    for (const auto& doc : docs) {
        for (const auto& w : doc) {
            if (w == term) {
                df++;
                break;
            }
        }
    }
    return std::log10((double)docs.size() / df) + 1;
}

int main() {
    std::vector<std::vector<std::string>> docs {
        { "リンゴ", "リンゴ" },
        { "リンゴ", "レモン" },
        { "レモン", "ミカン" }
    };
    std::vector<std::string> terms {
        "リンゴ", "レモン", "ミカン"
    };

    for (const auto& w : terms) {
        std::cout << "idf(" << w << ") = " << idf(w, docs) << std::endl;
    }
}
